crypto/src/merkle/mod.rs

```diff
@@ -127,6 +127,39 @@ impl<H: Hasher> MerkleTree<H> {
         Ok(MerkleTree { nodes, leaves })
     }
 
+    /// Returns new Merkle tree built from default, empty leaves using hash function
+    /// specified by the `H` generic parameter.
+    pub fn build_empty(depth: usize) -> Result<Self, MerkleTreeError> {
+        // Calculate number of leaves
+        let n = 2usize.pow(depth as u32);
+        // Build the leaves
+        let result = vec![0u8; n * std::mem::size_of::<H::Digest>()];
+
+        // translate a zero-filled vector of u8s into a vector of digests
+        let mut v = core::mem::ManuallyDrop::new(result);
+        let p = v.as_mut_ptr();
+        let len = n;
+        let cap = n;
+        let leaves = unsafe { Vec::from_raw_parts(p as *mut H::Digest, len, cap) };
+        // Determine hash values for nodes depending upon level
+        let mut cur_empty = leaves[0];
+        let mut empty_nodes = Vec::new();
+        for _ in 0..depth {
+            cur_empty = H::merge(&[cur_empty, cur_empty]);
+            empty_nodes.push(cur_empty);
+        }
+        // Begin with an empty internal node set
+        let mut nodes = vec![cur_empty];
```

Comment:
Same here, I would already allocate the whole space necessary for storing the internal nodes. Here we are reallocating at every power of 2.